### PR TITLE
[feature](meta-service) Support querying and adjusting rpc qps limit on meta service

### PR DESCRIPTION
**Pull request by Hastyshell**
## Proposed changes

Usage
1. adjust limit
```
curl http://ms_ip:ms_port/MetaService/http/v1/adjust_rate_limit?${params}
```
| Entry      | Description |
| ----------- | ----------- |
| param         | uint64 qps_limit  |
|behavior   | set qps_limit global default value |
|example|```curl http://ms_ip:ms_port/MetaService/http/v1/adjust_rate_limit?qps_limit=5000000```|

| Entry      | Description |
| ----------- | ----------- |
| param         | uint64 qps_limit, string rpc_name  |
|behavior   | set RPC specific qps_limit  |
|example|curl http://ms_ip:ms_port/MetaService/http/v1/adjust_rate_limit?qps_limit=5000000&rpc_name=get_cluster|

| Entry      | Description |
| ----------- | ----------- |
| param         | uint64 qps_limit, string rpc_name, string instance_id  |
|behavior   | set instance qps_limit for specific RPC  |
|example|```ccurl http://ms_ip:ms_port/MetaService/http/v1/adjust_rate_limit?qps_limit=5000000&rpc_name=get_cluster&instance_id="doris-0"```|

| Entry      | Description |
| ----------- | ----------- |
| param         | uint64 qps_limit, string instance_id  |
|behavior   | set global qps_limit for specific instance  |
|example|```curl http://ms_ip:ms_port/MetaService/http/v1/adjust_rate_limit?qps_limit=5000000&instance_id="doris-0"```|

2. query limit

| Entry      | Description |
| ----------- | ----------- |
| param         | none  |
|behavior   | query qps limit for all RPC interface |
|example|```curl http://ms_ip:ms_port/MetaService/http/v1/list_rate_limit```|

### CODE REVIEW

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by gavinchou** [comment]
Add some usage examples in description.

**Comment by TangSiyang2001** [comment]
run buildall

File: cloud/src/meta-service/meta_service_http.cpp [hunk 1]
```diff
@@ -331,6 +338,128 @@ static HttpResponse process_alter_iam(MetaServiceImpl* service, brpc::Controller
...
+    auto instance_id = std::string {http_query(uri, "instance_id")};
+
+    auto process_invalid_arguments = [&]() -> HttpResponse {
+        return http_json_reply(MetaServiceCode::INVALID_ARGUMENT,
+                               fmt::format("invalid argument: qps_limit(required)={}, "
+                                           "rpc_name(optional)={}, instance_id(optional)={}",
+                                           qps_limit_str, rpc_name, instance_id));
+    };
+
+    static auto parse_qps_limit =
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_http.cpp:354]: it seems that abuse using lambdas

File: cloud/src/meta-service/meta_service_http.cpp [hunk 2]
```diff
@@ -331,6 +338,128 @@ static HttpResponse process_alter_iam(MetaServiceImpl* service, brpc::Controller
...
+        });
+    };
+
+    auto set_instance_rpc_qps_limit = [&]() {
+        return process_set_qps_limit([&](int64_t qps_limit) {
+            return service->rate_limiter()->set_rate_limit(qps_limit, rpc_name, instance_id);
+        });
+    };
+
+    // for 8 element in true table of params, register processor cb
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_http.cpp:418]: this part is confusing

File: cloud/src/rate-limiter/rate_limiter.cpp [hunk 3]
```diff
@@ -41,36 +46,92 @@ void RateLimiter::init(google::protobuf::Service* service) {
...
+                LOG(WARNING) << "failed to parse max_qps_limit to rpc: " << rpc_name
                              << " config: " << v;
             }
         }
     }
+    return rpc_name_to_max_qps_limit;
+}
+
+template <typename Callable>
+void for_each_rpc_name(google::protobuf::Service* service, Callable cb) {
```
Comment (gavinchou) [inline cloud/src/rate-limiter/rate_limiter.cpp:60]: `rpc_name`s are fixed after starting the service, consider storing the names in a container which can be a member of MetaServiceImpl

File: cloud/src/rate-limiter/rate_limiter.cpp [hunk 4]
```diff
@@ -41,36 +46,92 @@ void RateLimiter::init(google::protobuf::Service* service) {
...
     auto it = limiters_.find(rpc_name);
     if (it == limiters_.end()) {
         return nullptr;
     }
     return it->second;
 }
 
+bool RateLimiter::set_rate_limit(int64_t qps_limit) {
+    std::lock_guard lock(mutex_);
+    auto filter = [this](const auto& kv) { return !rpc_with_specific_limit_.contains(kv.first); };
```
Comment (gavinchou) [inline cloud/src/rate-limiter/rate_limiter.cpp:94]: is it "filter out"?

File: cloud/src/rate-limiter/rate_limiter.cpp [hunk 5]
```diff
@@ -93,6 +154,35 @@ bool RpcRateLimiter::get_qps_token(const std::string& instance_id,
...
+    max_qps_limit_ = max_qps_limit;
+    auto filter = [this](const auto& kv) {
+        return !instance_with_specific_limit_.contains(kv.first);
+    };
+    for (auto& [k, v] : qps_limiter_ | std::views::filter(std::move(filter))) {
+        v->set_max_qps_limit(max_qps_limit);
+    }
+}
+
+bool RpcRateLimiter::set_max_qps_limit(int64_t max_qps_limit, const std::string& instance_id) {
```
Comment (gavinchou) [inline cloud/src/rate-limiter/rate_limiter.cpp:168]: "limit" already implies "max", "max" in the name seems redundant

File: cloud/src/rate-limiter/rate_limiter.h [hunk 6]
```diff
@@ -58,15 +81,24 @@ class RpcRateLimiter {
      */
     bool get_qps_token(const std::string& instance_id, std::function<int()>& get_bvar_qps);
 
-    // Todo: Recycle outdated instance_id
+    std::string_view rpc_name() const { return rpc_name_; }
```
Comment (gavinchou) [inline cloud/src/rate-limiter/rate_limiter.h:84]: add comment for every function: behavior, assumptions, params and return value

File: cloud/src/rate-limiter/rate_limiter.h [hunk 7]
```diff
@@ -35,12 +38,32 @@ class RateLimiter {
...
+    bool set_rate_limit(int64_t qps_limit, const std::string& rpc_name,
+                        const std::string& instance_id);
+
+    // set instance level rate limit globally, will influence settings for the same instance of specific rpc
+    bool set_instance_rate_limit(int64_t qps_limit, const std::string& instance_id);
+
 private:
     // rpc_name -> RpcRateLimiter
     std::unordered_map<std::string, std::shared_ptr<RpcRateLimiter>> limiters_;
+    std::unordered_set<std::string> rpc_with_specific_limit_;
```
Comment (gavinchou) [inline cloud/src/rate-limiter/rate_limiter.h:65]: add comment, it seems to be "This collection contains the RPCs that have had a limit set by the specified name"

File: cloud/src/rate-limiter/rate_limiter.h [hunk 8]
```diff
@@ -35,12 +38,32 @@ class RateLimiter {
 public:
     RateLimiter() = default;
     ~RateLimiter() = default;
+
     void init(google::protobuf::Service* service);
+
     std::shared_ptr<RpcRateLimiter> get_rpc_rate_limiter(const std::string& rpc_name);
 
+    void for_each_rpc_limiter(
```
Comment (gavinchou) [inline cloud/src/rate-limiter/rate_limiter.h:46]: add comment even though it seems simple, describe: behavior, assumptions, params and return value

File: cloud/src/rate-limiter/rate_limiter.h [hunk 9]
```diff
@@ -75,12 +107,17 @@ class RpcRateLimiter {
...
 
+    void for_each_qps_token(std::function<void(std::string_view, std::shared_ptr<QpsToken>)> cb);
+
+    // Todo: Recycle outdated instance_id
+
 private:
     bthread::Mutex mutex_;
     // instance_id -> QpsToken
     std::unordered_map<std::string, std::shared_ptr<QpsToken>> qps_limiter_;
+    std::unordered_set<std::string> instance_with_specific_limit_;
```
Comment (gavinchou) [inline cloud/src/rate-limiter/rate_limiter.h:118]: ditto

File: cloud/src/meta-service/meta_service_http.cpp [hunk 10]
```diff
@@ -331,6 +338,128 @@ static HttpResponse process_alter_iam(MetaServiceImpl* service, brpc::Controller
...
+                    fmt::format("param `qps_limit` is not a legal int64 type:{}", ex.what()));
+        }
+        if (qps_limit < 0) {
+            return http_json_reply(MetaServiceCode::INVALID_ARGUMENT,
+                                   "`qps_limit` should not be less than 0");
+        }
+        return qps_limit;
+    };
+
+    auto process_set_qps_limit = [&](std::function<bool(int64_t)> cb) -> HttpResponse {
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_http.cpp:372]: abuse lambdas, too many  indentions and hard to read

**Review by gavinchou — CHANGES_REQUESTED** [review]

File: cloud/src/rate-limiter/rate_limiter.cpp (hunk 3 above)
Comment (TangSiyang2001) [reply cloud/src/rate-limiter/rate_limiter.cpp:60]: There's no need to add a container, this function is only called once when doing init.

File: cloud/src/rate-limiter/rate_limiter.cpp (hunk 4 above)
Comment (TangSiyang2001) [reply cloud/src/rate-limiter/rate_limiter.cpp:94]: nope, just as it means literally

File: cloud/src/rate-limiter/rate_limiter.cpp (hunk 5 above)
Comment (TangSiyang2001) [reply cloud/src/rate-limiter/rate_limiter.cpp:168]: just inheritate the original naming, will be mod

File: cloud/src/meta-service/meta_service_http.cpp (hunk 2 above)
Comment (TangSiyang2001) [reply cloud/src/meta-service/meta_service_http.cpp:418]: Maybe I'll add some comments, just register different callback for different params pattern, and the "abuse" of lamda are just meant to adapt for this.

**Review by luwei16 — APPROVED** [review]

**Comment by TangSiyang2001** [comment]
run buildall

**Comment by TangSiyang2001** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Comment by TangSiyang2001** [comment]
run compile

**Comment by TangSiyang2001** [comment]
run cloud_p0

**Comment by TangSiyang2001** [comment]
run p0

**Comment by TangSiyang2001** [comment]
run external

**Pull request merged**